nexus/src/authz/roles.rs

```diff
@@ -0,0 +1,206 @@
+    /// can affect access to this resource, return the parent resource.
+    /// Otherwise, returns `None`.
+    fn parent(&self) -> Option<Box<dyn AuthzResource>>;
+}
+
+impl<T: AuthzApiResource> AuthzResource for T {
+    fn fetch_all_related_roles_for_user<'a, 'b, 'c, 'd, 'e, 'f>(
+        &'a self,
+        opctx: &'b OpContext,
+        datastore: &'c DataStore,
+        authn: &'d authn::Context,
+        roleset: &'e mut RoleSet,
+    ) -> BoxFuture<'f, Result<(), Error>>
+    where
+        'a: 'f,
+        'b: 'f,
+        'c: 'f,
+        'd: 'f,
+        'e: 'f,
+    {
+        async move {
+            // If the user is authenticated ...
+            if let Some(actor_id) = authn.actor() {
+                // ... and if roles can be assigned directly on this resource ...
+                if let Some((resource_type, resource_id)) = self.db_resource() {
+                    // ... then start by fetching all the roles for this user
+                    // that are associated with this resource.
+                    trace!(opctx.log, "loading roles";
+                        "actor_id" => actor_id.0.to_string(),
+                        "resource_type" => ?resource_type,
+                        "resource_id" => resource_id.to_string(),
+                    );
+                    let roles = datastore
+                        .role_asgn_builtin_list_for(
+                            opctx,
+                            actor_id.0,
+                            resource_type,
+                            resource_id,
+                        )
+                        .await?;
+                    // Add each role to the output roleset.
+                    for role_asgn in roles {
+                        assert_eq!(
+                            resource_type.to_string(),
+                            role_asgn.resource_type
+                        );
+                        roleset.insert(
+                            resource_type,
+                            resource_id,
+                            &role_asgn.role_name,
+                        );
+                    }
+                }
+
+                // If this resource has a parent, the user's roles on the parent
+                // might grant them access to this resource.  We have to fetch
+                // those, too.  This process is recursive up to the root.
+                //
+                // (In general, there could be another resource with _any_ kind
+                // of relationship to this one that grants them a role that
+                // grants access to this resource.  In practice, we only use
+                // "parent", and it's clearer to just call this "parent" than
+                // "related_resources_whose_roles_might_grant_access_to_this".)
+                if let Some(parent) = self.parent() {
+                    parent
+                        .fetch_all_related_roles_for_user(
+                            opctx, datastore, authn, roleset,
```

Comment:
Doesn't need to be addressed in this PR, but since we have the whole chain of `(ID, resource type)` in memory here, we should be able to pull them out into a list and have `role_asgn_builtin_list_for` query for them all at once in a single query.

Reply:
Yeah, it might be a nice improvement if this function, rather than making the query and updating `RoleSet`, actually just stored the relevant (resource_type, id) tuples into a BTreeSet, and then we used that at the end to build a sort of monstrous query in `role_asgn_builtin_list_for`.